### ISSUE TITLE
Linux 6.7 super_block has s_shrink failing

### ISSUE THREAD

**Issue by ptr1337** [openzfs/zfs#15518]
Kernel: 6.7 rc1
Operating System: CachyOS (Archlinux)

Relevant Log:

```
checking whether sops->dirty_inode() wants flags... yes
checking whether super_block has s_shrink... configure: error: 
        *** None of the expected "sb->s_shrink()" interfaces were detected.
        *** This may be because your kernel version is newer than what is
        *** supported, or you are using a patched custom kernel with
        *** incompatible modifications.
        ***
        *** ZFS Version: zfs-2.2.99-174_g043c6ee3b
        *** Compatible Kernels: 3.10 - 6.5
```

**Comment by satmandu** [comment]
This might be due to the lockless slab shrink work?

e.g. https://lore.kernel.org/linux-erofs/20230724094354.90817-41-zhengqi.arch@bytedance.com/
https://lwn.net/Articles/940644/

**Comment by ascendbeing** [comment]
I believe it can be narrowed down to this commit: https://github.com/torvalds/linux/commit/ecae0bd5173b1014f95a14a8dfbe40ec10367dcf
it looks like they changed the s_shrink format in this commit to something zfs does not currently anticipate.
so maybe try building against the commit just prior to this one, or maybe someone can do a patch/PR
I was looking at https://github.com/torvalds/linux/commits/master/include/linux/fs.h to find this.

**Comment by ckane** [comment]
Thanks for all the details, I saw this too on my Arch system when trying to build the new kernel, but haven't had a chance to dig in.

**Comment by spikerguy** [comment]
Any update on this ? 

I tried to build on aarch64 with upstream kernel 6.7-rc3 and failed with same error. 

Thanks.

**Comment by setar** [comment]
kernel 6.7-rc5 same error.

**Comment by DominoTree** [comment]
Just hit this with a Fedora update - I suspect others will be hitting it soon as well

**Comment by satmandu** [comment]
Wait for OpenZFS 2.2.3 to come out...

**Comment by DominoTree** [comment]
> For the Fedora 39 users out there that need to roll back the kernel to get around this...
> 
> ```
> dnf install kernel-\*-6.5.6
> dnf (re)install zfs zfs-dkms
> ```

What I did was the following, to get a list of installed kernels and use an existing one prior to 6.7.x (6.6.14 happened to be the most recent):

```
grubby --info=ALL
grubby --set-default=/boot/vmlinuz-6.6.14-200.fc39.x86_64
```

**Comment by edgan** [comment]
I took the 6.7 commits from this PR, https://github.com/openzfs/zfs/pull/15860, and with a little tweaking made a usable patch for ZFS 2.2.2 and Linux kernel 6.7.x. I have successfully used it to compile modules for 6.7.3 and 6.7.4. I am currently running it with 6.7.4 on Fedora 39.
  
  **Use at your own risk.**
  
Steps:
```
cd /usr/src
wget https://cygnusx-1.org/zfs-2.2.2-6.7.patch
cd zfs-2.2.2
cat ../zfs-2.2.2-6.7.patch | patch -p1
dkms build -m zfs -v 2.2.2 -k 6.7.3-200.fc39.x86_64
dkms install -m zfs -v 2.2.2 -k 6.7.3-200.fc39.x86_64
```

**Comment by gpmidi** [comment]
@tleydxdy The older one is probably installed to meet deps; no guarantee about which runs though.

**Comment by tleydxdy** [comment]
right I mean it should block 6.7 since that is not in even in the supported range.

**Comment by DominoTree** [comment]
> echo 'zfs' > /etc/dnf/protected.d/zfs.conf does the trick

Aha! I think this is the 'right' way to resolve this issue until the new version is released.

**Comment by DominoTree** [comment]
> echo 'zfs' > /etc/dnf/protected.d/zfs.conf does the trick

This edit will prevent DNF from *uninstalling* ZFS while doing other updates, but won't prevent DNF from upgrading to a kernel that makes the DKMS module fail to build - ask me how I know :P

**Comment by tleydxdy** [comment]
exactly, the zfs packages should block kernels with unsupported major version from even installing, it still might break for minor version updates I suppose but it's a lot better than nothing

**Comment by YesThatGy** [comment]
Fedora 39 user here. Ran into this a few weeks ago, did an update and still not fixed. 

 "the zfs packages should block kernels with unsupported major version from even installing" 
 
 Can I +100 vote for this?

**Comment by DominoTree** [comment]
I wonder if it’s possible to trigger akmods pointed at the new kernel during the dnf update process, so that the transaction is automatically rolled back if the dkms module fails to build against the new kernel.If not, I think the zfs dkms module would have to be built against kernels as they hit staging (or similar) and the release zfs package metadata would have to be updated if a failure occurs.On Feb 20, 2024, at 9:43 PM, YesThatGy ***@***.***> wrote:﻿
Fedora 39 user here. Ran into this a few weeks ago, did an update and still not fixed.
"the zfs packages should block kernels with unsupported major version from even installing"
Can I +100 vote for this?

**Comment by tleydxdy** [comment]
I think that is a bit overkill, at the most basic, say zfs-2.2.2 is released with "Linux: compatible with 3.10 - 6.6 kernels" then the package can simply be mark as incompatible with kernel <=3.8 and kernel >=6.7

**Comment by AllKind** [comment]
dkms does not provide any mechanism to talk to each possible packet manager out there (AFAIK).
If a distro is packaging the zfs-dkms or kmods it is responsible for configuring the relevant 'Conflict' or whatever definitions for their package (.deb or .rpm for example).
Now the packages built from source by OpenZFS would need to identify each distribution (i.e. fedora, ubuntu, etc), most likely in the ./configure process and pass that information to the package building scripts (for i.e. .deb or .rpm) and there depending on which distro is in play define the 'Conflict' with the correct package name for that distribution (i.e. linux-source, kernel-source, kernel-devel, etc.)
See my comment: https://github.com/openzfs/zfs/issues/15759#issuecomment-1951420427
Nobody has done the work (yet).
I bet the OpenZFS developers would more then welcome a Pull Request from somebody volunteering for that.

**Comment by robn** [comment]
FYI, 2.2.3 is released: https://github.com/openzfs/zfs/releases/tag/zfs-2.2.3

**Comment by DominoTree** [comment]
Currently traveling but I think I'll take a crack at it when I get back next week - seems like a fairly easy area to get involved without knowing a ton about ZFS internals :)

**Comment by AllKind** [comment]
@DominoTree 
Maybe this will come in handy: https://fedoraproject.org/wiki/Packaging:Conflicts

**Comment by AllKind** [comment]
@DominoTree 
I didn't notice it before, but looked at the zfs-dkms.spec.in file now and I see the min/max version and a requires definition are already there:

```
Requires:       kernel-devel >= @ZFS_META_KVER_MIN@, kernel-devel <= @ZFS_META_KVER_MAX@.999
Requires(post): kernel-devel >= @ZFS_META_KVER_MIN@, kernel-devel <= @ZFS_META_KVER_MAX@.999
```

So either that is not sufficient for RPM (Do we need a Conflicts definition?), or something else in the chain is broken??

**Comment by ptr1337** [comment]
Ill close this bug, since it is fixed in upstream.

**Comment by tleydxdy** [comment]
@AllKind @ptr1337 yes, require is not enough, need to have conflict as well. that line is already there but me and a few others still get broken system from kernel upgrades. also kernel-devel is different from kernel, it's only the headers

**Comment by ptr1337** [comment]
> @AllKind @ptr1337 yes, require is not enough, need to have conflict as well. that line is already there but me and a few others still get broken system from kernel upgrades. also kernel-devel is different from kernel, it's only the headers

I think this should be in a separate issue discussed since this is not a direct issue with the current 6.7 issue described above.

**Comment by poige** [comment]
> Ill close this bug, since it is fixed in upstream.

@ptr1337 That's ambigous — which upstream(?), if there're 2 involved projects (kernel and ZFS).

References would be handy.

```
checking whether super_block has s_shrink... configure: error:
        *** None of the expected "sb->s_shrink()" interfaces were detected.
[…skip…]
        *** ZFS Version: zfs-2.2.3-1
        *** Compatible Kernels: 3.10 - 6.7
```
— I'm facing this very issue with zfs-2.2.3-1 and 6.7.9

**Comment by AllKind** [comment]
@poige 
ptr1337 meant ZFS upstream. Namely by this pull request: https://github.com/openzfs/zfs/pull/15694
If it fails for you with kernel 6.7, you should open a new issue providing the details of your setup.
You should also provide your `config.log`.

**Comment by robn** [comment]
2.2.3 builds fine here against 6.7.9. More information required. I suggest opening a new issue; this one is closed and no one is watching it closely.

**Comment by poige** [comment]
Thanks for info both @robn and
@AllKind … as to what I should do — I'm sure everybody should mind theirs own business and avoid telling people what they should do. ;-P

**Comment by AllKind** [comment]
@poige 

> as to what I should do — I'm sure everybody should mind theirs own business and avoid telling people what they should do. ;-P

WTF?
I was only trying to be of assistance. As robn said, this is a closed thread. Doesn't get much attention.
I didn't say you **must** do this, or that. I was only advising what would be most helpful to you and the developers in order to help **you!**
How that could be interpreted as patronizing is beyond me.
If I would mind my own business, I wouldn't try to help you!
